extension/src/fileSystem/index.ts

```diff
@@ -200,14 +200,60 @@ stages:
   return appendFileSync(dvcYamlPath, pipeline)
 }
 
+const loadYamlAsDoc = (
+  path: string
+): { doc: yaml.Document; lineCounter: yaml.LineCounter } | undefined => {
+  try {
+    const lineCounter = new yaml.LineCounter()
+    return {
+      doc: yaml.parseDocument(readFileSync(path, 'utf8'), { lineCounter }),
+      lineCounter
+    }
+  } catch {
+    Logger.error(`failed to load yaml ${path}`)
+  }
+}
+
+const getYamlFileIndent = (lines: string[]) => {
+  const firstLineWithInd = lines.find(line => line.startsWith(' ')) || ''
+  const spacesMatches = firstLineWithInd.match(/^( +)[^ ]/)
+  const spaces = spacesMatches?.[1]
+
+  return spaces ? spaces.length : 2
+}
+
 export const addPlotToDvcYamlFile = (cwd: string, plotObj: PlotConfigData) => {
   const dvcYamlFile = `${cwd}/dvc.yaml`
+  const dvcYamlDoc = loadYamlAsDoc(dvcYamlFile)
+
+  if (!dvcYamlDoc) {
+    return
+  }
+
+  const { doc, lineCounter } = dvcYamlDoc
   const { dataFile, ...plot } = plotObj
   const plotName = relative(cwd, dataFile)
   const dvcYamlLines = readFileSync(dvcYamlFile, 'utf8').split('\n')
-  const plotYaml = yaml.stringify({ plots: [{ [plotName]: plot }] }).split('\n')
+  const indent = getYamlFileIndent(dvcYamlLines)
+
+  const plots = doc.get('plots', true) as yaml.YAMLSeq | undefined
```

Comment:
We use the `yaml` package's [Document API](https://eemeli.org/yaml/#documents) to check if the plots exist and get the last position of the plots item for inserting lines into the yaml file.

Reply:
As a follow up can you PTAL at `vscode`'s `WorkspaceEdit` API (should be able to use `insert`). We don't have to use it but it could simplify things.